### PR TITLE
feat(jdbc): stringify nested types

### PR DESCRIPTION
**Pull request by Mause**
It looks like DBeaver is smart enough to parse lists into a format it can decompose, without any further work, though maps and structs are left as is.

This will probably fulfill most people's requirements for nested types in the JDBC driver

Screenshot_2023-04-08_00-24-40

### CODE REVIEW

File: tools/jdbc/src/jni/duckdb_java.cpp [hunk 1]
```diff
@@ -752,6 +752,10 @@ JNIEXPORT jobjectArray JNICALL Java_org_duckdb_DuckDBNative_duckdb_1jdbc_1fetch(
 			break;
 		case LogicalTypeId::TIME:
 		case LogicalTypeId::DATE:
+		case LogicalTypeId::STRUCT:
+		case LogicalTypeId::UNION:
+		case LogicalTypeId::LIST:
+		case LogicalTypeId::MAP:
```
Comment (Mytherin) [inline tools/jdbc/src/jni/duckdb_java.cpp:758]: Perhaps we should do this for all unsupported types?

File: tools/jdbc/src/jni/duckdb_java.cpp [hunk 2]
```diff
@@ -752,6 +752,10 @@ JNIEXPORT jobjectArray JNICALL Java_org_duckdb_DuckDBNative_duckdb_1jdbc_1fetch(
...
 		case LogicalTypeId::INTERVAL: {
 			Vector string_vec(LogicalType::VARCHAR);
 			VectorOperations::DefaultCast(vec, string_vec, row_count);
```
Comment (Mytherin) [inline tools/jdbc/src/jni/duckdb_java.cpp:761]: `DefaultCast` here means that extension types will not be correctly converted to string - if we could change this to a `VectorOperations::Cast` instead we can convert those correctly.

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Great to get this working, and nice that DBeaver handles this smoothly even as strings. Some comments:

**Comment by Mytherin** [comment]
Thanks for the fixes! LGTM

**Comment by Jack-Corbett** [comment]
Excellent work @Mause can't thank you enough for fixing this! @Mytherin do you know when this will make it to release?

**Comment by Mytherin** [comment]
The next release is planned for mid May

**Comment by lukaseder** [comment]
As I've commented in https://github.com/duckdb/duckdb/discussions/7173, I think that the current next-version `String` representation of arrays is going to be insufficient for formal tool support. Even Dbeaver can't distinguish between `array['a', 'b']` and `array['a, b']`, so more formal JDBC `Array` and `Struct` support would be helpful here.

image

**Pull request merged**